Add architecture part in the Whitepaper (#237)
okp4/docs: docs/whitepaper/architecture.mdx

---
sidebar_position: 4
---

# Architecture

## General overview

The protocol provides a disruptive access control mechanism. Whereas a central gateway usually safekeeps a specific set of resources, OKP4 protocol acts as a global source of trust for any operation:

![OKP4 Basic Architecture Overview](/img/content/whitepaper/arch-basic.webp)

Through an application form, providers indicate the rules for consuming its resources. The interface interacts with the blockchain, which stores the resources descriptions, revenue conditions, restrictions and execution instructions.

> A consumer asks the blockchain through an app to execute a request using multiple resources. If the rules set by the providers are met, the protocol emits an event to validate the request.

Reviewer: and the zone?

### Built on Cosmos

The OKP4 blockchain is built upon the Cosmos ecosystem to benefit from advanced implementation options without reinventing the wheel. Cosmos's modular architecture enables customized application-specific chains with ease. This unique flexibility fits the project's requirements, ensuring optimal performance and resource allocation.

Moreover, with Cosmos's Inter-Blockchain Communication (IBC) protocol, the OKP4 blockchain can communicate and transact seamlessly with other chains within the Cosmos ecosystem and external networks. This interchain compatibility expands the reach and potential user base and opens avenues for collaboration and synergy with other innovative projects in the Cosmos ecosystem.

## Modules

The modules implement core functions particular to the OKP4 blockchain. They provide features for protocol governance, rules interpretation regarding resources, and more.

### Logic

> Any smart contract deployed on the OKP4 blockchain uses the logic module to evaluate sharing conditions definitions and user access queries written in Prolog. Prolog is a powerful declarative programming language to manage from simple restrictions to complex rules. This language is not natively operable on-chain; the logic module adds Prolog interpretation capability.

Author: I forgot "can". 
Updated: 
"Thus, any smart contract deployed on the OKP4 blockchain can use the logic module to evaluate queries written in Prolog."

The logic module includes custom predicates, like blockchain state clauses and various data operation utilities, enabling advanced governance possibilities.

> The logic module has a virtual file system built in. Users can store Prolog programs and add a reference to another Prolog program. It enhances reusability and composability.  

Author: OK I deleted these two sentences

Two of them have a custom implementation: the mint and the vesting modules.
“Mint”, used for creating new $KNOW units, applies specific tokenomics parameters. It determines the inflation rate for validators’ and stackers’ earnings.
“Vesting” refers to distributing tokens over a specified period rather than all at once. This approach allows for controlled token distribution over time, promoting responsible token management and preventing immediate sell-offs that could negatively impact the OKP4 ecosystem.

### WASM

OKP4 incorporates the CosmWasm smart contracting platform built for the Cosmos Ecosystem. The primary programming language used in this module is Rust for building secure and multi-chain smart contracts. Yet, any language compiled into WASM can be supported as they become available.

## Smart Contracts

### Smart contract overview

Smart contracts handle all interactions with external parties, whether to provide a digital resource with condition rules to use it, to consume this resource, or to send payments. These decentralized programs are centered around three fundamental pillars:

> * Ontology: the knowledge graph modelling all the different resources (e. g. datasets, computation services, infrastructure, orchestration engines, …), their descriptive properties and relations between each other. It also includes access rules and execution instructions linked to resources. The ontology is a whole of textual objects, a way to store a semantic representation of the Dataverse.

Author: ✔️

* Governance: the policies ruling the definition and the validation of resource permissions.  
Rules (i.e. the conditions to give access to resources) can be applied at one among multiple levels: for everyone, for a set of resources, for a specific resource or between selected parties only.

> * Orchestration: the processes that check the access conditions compliance defined by entities. Orchestration relates to resource authorisation controls and revenue-sharing distribution.

Author: Updated: 

### Smart contract overview

Smart contracts handle all interactions with external parties, whether to provide a digital resource with condition rules to use it, to consume this resource, or to send payments. These decentralized programs are centered around three fundamental pillars:

* **Ontology**: models all the different resources (e. g. datasets, computation services, infrastructure, orchestration engines, …), their descriptive properties and relations between each other.  
It also includes access rules and execution instructions linked to resources. The ontology is a whole of objects, a way to store a semantic representation of the Dataverse.
* **Governance**: statutes the definition and the validation of resource permissions.  
Rules (i.e. the conditions to give access to resources) can be applied at one among multiple levels: for everyone, for a set of resources, for a specific resource or between selected parties only.
* **Orchestration**: ensures the smooth coordination and execution of services.  
It enforces the rules and policies that govern resource (datasets and services) authorization.

<center><img src="/img/content/whitepaper/arch-3pillars.webp"></img></center>

### Objectarium: unstructured data storage

> Prolog code defines governance rules, the “law”. These pieces of logic programming (usually from `.pl` files) are stored in raw with an `objectarium` smart contract, which acts as a file system.

Author: Updated: 

### Objectarium: unstructured data storage

The main goal of `objectarium` is to provide the protocol with a versatile mechanism for storing unstructured data, ensuring that the stored data remains immutable.

Any user or smart contract can ”pin” a stored object. Object deletion is impossible as long as it contains at least one “pin”. The user (or smart contract) should use the “unpin” function from `objectarium` to remove a pin.

A specific use case is the storing governance rules as a Prolog program. Prolog code defines governance rules, the “law”. These pieces of logic programming (usually from `.pl` files) are stored in raw with an `objectarium` smart contract.

> Prolog code defines governance rules, the “law”. These pieces of logic programming (usually from `.pl` files) are stored in raw with an `objectarium` smart contract, which acts as a file system.

Author: I agree. We should check (in code comments, ...) to remove any "file system" reference for consistency.

Any user or smart contract can ”pin” a stored object. Object deletion is impossible as long as it contains at least one “pin”. The user (or smart contract) should use the “unpin” function from `objectarium` to remove a pin.

### Cognitarium: semantic data storage

The external resources description, usage rules and execution instructions are defined with an ontology, specifying metadata and relationships.

> There is no restriction on the structure to allow ample expressiveness. But the OKP4 “core”, the minimal ontology scheme, requires the specification of an explicit relationship between an external resource, its governance rules and service execution instructions.

Author: Thanks, I didn't succeed in putting words on it.

### Law-stone: rules agent

To create a governance rule, one should use the `law-stone` smart contract and instantiate it with a Prolog code.

A `law-stone` instance stores the Prolog code using `objectarium` smart-contract, pins it, and checks the dependencies to pin. Indeed, if the provided Prolog code depends on another Prolog program (already stored with `objectarium`), it’s necessary to ensure its availability.

The rule cannot be changed; what is stored onchain is immutable. The one who instantiates the governance rule is the administrator, the only one able to destroy it. When a law is deleted, the main Prolog code stored in `objectarium` is unpinned. All object with no other pin is deleted.

The `law-stone` has an “ask” method to execute Prolog queries from stored rules. It loads them from `objectarium`, and uses the logic module to get the answer with eventual substitutions for results.

### Pactum: service agreement contract

> This smart contract guarantees the orchestration, the respect of the law and the value allocations.

Author: ✔️

A `pactum` instantiation defines a service agreement between at least two parties, details the prerequisites that fulfill its enforcement and the actions to proceed accordingly. These conditions to execute contractual obligations are materialized as a set of terms previously stored with the `law-stone` smart contract.  

> A `pactum` instance records service execution states and applies token revenue-sharing distribution if conditions are met. It can also flag invalid service execution and block escrow in consequence.

Author: Updated: not quite sure. Is it OK?

### Pactum: service agreement contract

This smart contract guarantees the orchestration, the respect of the law and the value allocations.

A `pactum` instantiation defines an agreement between at least two parties and details the prerequisites that fulfil its enforcement and the actions to proceed accordingly. These conditions to execute contractual obligations are materialized as a set of terms previously stored with the `law-stone` smart contract.  

Service execution states are recorded in the ontology, forming the knowledge graph in `cognitarium`. The evaluation of conditions that must be met within the agreement changes over time based on the "facts" are also recorded in the ontology. Thus, a `pactum` instance applies token revenue-sharing distribution if conditions are met. It can also flag invalid service execution and block escrow in consequence.

`zone-hub` is the unique interface with the off-chain interfaces for any blockchain interaction. This smart contract operates external requests and calls the other smart contracts methods. It checks errors and ensures authentication for protected operations.  

When a service provider submits an entry to add new resources policies, `zone-hub` stores the Prolog code with an instantiation of a `law-stone` smart contract. 

`zone-hub` populates the ontology with `law-stone` address as a rules references, service execution instructions and all relevant resource metadata, via an `InsertData` message of a `cognitarium` contract. This process applies to an agreement between parties, a resource consent or a zone (resources categorization) definition.

When a consumer submits a query, the `zone-hub` instance evaluates the required resources with a request to `cognitarium`. It extracts the governance rules, and, to evaluate resource access, it asks the related `law-stone` instance the Prolog interpretation. If the response back is positive, `zone-hub` validates the user transaction and emits an event. This event triggers a service execution to (off-chain) orchestration services.

> `zone-hub` also handles state reportings from orchestration executions. It redirects messages to `pactum` instances.

Reviewer 2: Nope. See previous comment. It stores this to the `cognitarium` (after validating the data provided).

Author: Updated:

`zone-hub` also handles state reportings from orchestration executions. It stores messages in a `cognitarium` instances, before an evaluation by a `pactum` instance.

### End user interfaces

Newly generated knowledge could be presented with customised dashboards from any business intelligence component.
White-label data platforms graph the results obtained by running workflows with the OKP4 protocol. 

Since some companies cannot use a wallet to query the protocol, end user interfaces might also provides a layer of abstraction to bring data on request without any blockchain interaction.

## Connect’em all

Let’s recap with a concrete example applied to the AI industry.

Two companies (Corp A and Corp B) grant access to their databases, while a third company (Corp C)  provides a machine-learning training workflow using the data. A data scientist (Individual D) wants to get new knowledge with resources from Corp A, Corp B and Corp C.
The OKP4 solution orchestrates the training process without exposing raw data, ensuring privacy, sovereignty and security. Moreover, the protocol provides revenue-sharing conditions and immutable records of the ML model's training sources.

> 1. Each on their own, Corp A and Corp B indicate through an administration portal the underlying technology of their respective database. They store access tokens with a Secret Management Service, with authorization for services from Corp C.

Reviewer 2: This is not clear to me. As I understand it, service C accesses the authentication key to gain authorization for data consumption. However, for both protocol mediation and security mediation purposes, access to this key should only be possible by the Orchestration service's identity alone within the context of a properly registered execution order within the protocol.

Author: I initially differentiated the service/algo provided by C from the orchestration service as two different "actors". Then I realized, especially with the sequence diagram, that it was overloading the whole. 

What would you suggest: make a clear difference between these two services, or explain with a phrase that we combine to simplify, to make it more readable?

3. Metadata indicate that resources shared by Corp A and Corp B are compatible with a ML job from Corp C. Corp C also uses the administration portal to submit its training model algorithm, governance rules, and service execution instructions.

4. Individual D parameters its cloud environment with the portal, setting up how an orchestration service should store the execution request results.

5. Individual D uses the portal to request service execution with the ML Workflow from Corp C, using data from Corp A and Corp B. He submits a transaction with a Keplr wallet and pays with $KNOW tokens. The blockchain validates the execution request (access and execution authorizations).

6. The orchestration service listens to the event from the blockchain. It recovers access keys, and then executes the workflow training algorithm from Corp C, using Corp A & Corp B data it fetches.

> 7. The orchestration service reports the job state changes. If all works well, service agreement rules are applied, $KNOW tokens from Individual D are unlocked to Corp A, Corp B et Corp C.

Reviewer 2: ```suggestion
7. The orchestration service tracks the progress and state changes of all jobs within the workflow and reports this information to the blockchain. If all works well, service agreement rules are applied, $KNOW tokens from Individual D are unlocked to Corp A, Corp B et Corp C.
```

---
sidebar_position: 4
---

# Architecture

## General overview

The protocol provides a disruptive access control mechanism. Whereas a central gateway usually safekeeps a specific set of resources, OKP4 protocol acts as a global source of trust for any operation:

> ![OKP4 Basic Architecture Overview](/img/content/whitepaper/arch-basic.webp)

Reviewer 3: I'm not sure about the 4th step definition in the diagram: `report usage`.

It is more than a usage, it's a status. It shall contains any relevant information regarding execution, not only usage.

Through an application form, providers indicates the rules for consuming its resources. The interface interacts with the blockchain, which stores the resources descriptions, revenue conditions, restrictions and execution instructions.

> A consumer asks the blockchain through an app to compute a query using multiple resources. If the rules set by the providers are met, the protocol emits an event to validate the query.

Reviewer 3: The term `query` isn't exact as it refers usually to read operations, the intent here is more an `execution request`. The diagram above contains references to this term also.

An orchestration service, the guard on a resource relying solely on blockchain validation events, processes the query from the consumer. This query creates new knowledge as it leverages multiple resources from different providers with no required trust on a third party.
The orchestration service reports the execution states to the blockchain to ensure reliable logging and payment.

Smart contracts carry all these interactions with the blockchain. Nodes validate smart contracts execution and persistent storage through a consensus. A distributed database, the so-called blockchain, commits to every change. The OKP4 blockchain is built with custom modules enabling the protocol governance and validation rules capabilities, among other functionalities.

The diagram above provides a simplified explanation of the system’s overall operation. More precisely, the entire solution architecture is structured as follows:

> ![OKP4 Architecture Schema](/img/content/whitepaper/arch-overview.webp)

Reviewer 3: The docker icon on the workflow engine bothers me, it's incompatible with the _"Anything as a Service"_ approach.

The arrows coming out of the blockchain are problematic for me as it suggests an action from the blockchain onto the external world. I understand this diagram is an overview and not precise, but this is false.

Two of them have a custom implementation: the mint and the vesting modules.
“Mint”, used for creating new $KNOW units, applies specific tokenomics parameters. It determines the inflation rate for validators’ and stackers’ earnings.
“Vesting” refers to distributing tokens over a specified period rather than all at once. This approach allows for controlled token distribution over time, promoting responsible token management and preventing immediate sell-offs that could negatively impact the OKP4 ecosystem.

### WASM

OKP4 incorporates the [CosmWasm](https://cosmwasm.com/) smart contracting platform built for the Cosmos Ecosystem. The primary programming language used in this module is [Rust](https://www.rust-lang.org/) for building secure and multi-chain smart contracts. Yet, any language compiled into WASM can be supported as they become available.

## Smart Contracts

### Smart contract overview

Smart contracts handle all interactions with external parties, whether to provide a digital resource with condition rules to use it, to consume this resource, or to send payments. These decentralized programs are centered around three fundamental pillars:

> * **Ontology**: models all the different resources (e. g. datasets, computation services, infrastructure, orchestration engines, …), their descriptive properties and relations between each other.  

Reviewer 4: ```suggestion
* **Ontology**: models all the different resources (e. g. datasets, computation services, infrastructure, orchestration engines, …), their descriptive properties and relations with each other.  
```

<center><img src="/img/content/whitepaper/arch-3pillars.webp"></img></center>

### Objectarium: unstructured data storage

The main goal of `objectarium` is to provide the protocol with a versatile mechanism for storing unstructured data, ensuring that the stored data remains immutable.

Any user or smart contract can ”pin” a stored object. Object deletion is impossible as long as it contains at least one “pin”. The user (or smart contract) should use the “unpin” function from `objectarium` to remove a pin.

A specific use case is the storing governance rules as a Prolog program. Prolog code defines governance rules, the “law”. These pieces of logic programming (usually from `.pl` files) are stored in raw with an `objectarium` smart contract.

### Cognitarium: semantic data storage

The external resources description, usage rules and execution instructions are defined with an ontology, specifying metadata and relationships.

There is no restriction on the structure to allow ample expressiveness. But the `core` serves as a fundamental and essential structure that defines concepts and their relationships, forming the basis upon which additional extensions can be integrated.

> Once inserted via the `cognitarium` smart contract, one can fetch the data from a “select” or a “describe” query. In the first case, it returns resources matching the criterias defined by the provided query. The second case returns the raw part of the ontology, the resource description identified by an Internationalized Resource Identifier (IRI).

Reviewer 4: ```suggestion
Once inserted via the `cognitarium` smart contract, one can fetch the data from a “select” or a “describe” query. In the first case, it returns resources matching the criteria defined by the provided query. The second case returns the raw part of the ontology, the resource description identified by an Internationalized Resource Identifier (IRI).
```

The `law-stone` has an “ask” method to execute Prolog queries from stored rules. It loads them from `objectarium`, and uses the logic module to get the answer with eventual substitutions for results.

### Pactum: managing agreements

This smart contract guarantees the orchestration, the respect of the law and the value allocations. It can also handle more aspects regarding agreements, such as escrow mechanisms.  

A `pactum` instantiation defines an agreement between at least two parties and details the prerequisites that fulfil its enforcement and the actions to proceed accordingly. These conditions to execute contractual obligations are materialized as a set of terms previously stored with the `law-stone` smart contract.  

Service execution states are recorded in the ontology, forming the knowledge graph in `cognitarium`. The evaluation of conditions that must be met within the agreement changes over time based on the "facts" are also recorded in the ontology. Thus, a `pactum` instance applies token revenue-sharing distribution if conditions are met. It can also flag invalid service execution and block escrow in consequence.

### Zone hub: entry point

> `zone-hub` is the unique interface with the off-chain interfaces for any blockchain interaction. This smart contract operates external requests and calls the other smart contracts methods. It checks errors and ensures authentication for protected operations.  

Reviewer 3: I would say that it's the single entrypoint to any mutations in the protocol, for example nothing prevents to query the `cognitarium` or a `law-stone` directly.

Regarding the term `authentication`, I think a proper word would be `authorization`.

When a service provider submits an entry to add new resources policies, `zone-hub` stores the Prolog code with an instantiation of a `law-stone` smart contract. 

`zone-hub` populates the ontology with `law-stone` address as a rules references, service execution instructions and all relevant resource metadata, via an `InsertData` message of a `cognitarium` contract. This process applies to an agreement between parties, a resource consent or a zone (resources categorization) definition.

> When a consumer submits a query, the `zone-hub` instance evaluates the required resources with a request to `cognitarium`. It extracts the governance rules, and, to evaluate resource access, it asks the related `law-stone` instance the Prolog interpretation. If the response back is positive, `zone-hub` validates the user transaction and emits an event. This event triggers a service execution to (off-chain) orchestration services.

Reviewer 3: Same as above regarding the `query` term :)

`zone-hub` also handles state reportings from orchestration executions. It stores messages in a `cognitarium` instances, before an evaluation by a `pactum` instance. 

## Orchestration

Orchestration services have access to off-chain resources. They only trust the OKP4 blockchain to execute jobs.

> Initially, there will be a unique orchestration service, a central “workflow engine”.  

Reviewer 3: Yes, an orchestration service should be able to access other off-chain resources in order to fulfill its task, executing a workflow. But the OKP4 blockchain doesn't execute jobs, and initially, who says there will be a unique orchestration service? 👀

SSI allows individuals or entities to control their own digital identities, while DIDs provide a unique and persistent identifier for these entities.

### Access authorization with Secret Management Services

Service providers should submit API keys (or private access instructions) to a secure storage solution and grant a read operation for an orchestration service. By utilizing SSI and DIDs, a Secret Management Service relies on the cryptographic integrity of the DIDs and verifiable credentials.

Service providers must add service execution instructions regarding secrets keys alongside the governance rules when referencing a resource to the on-chain ontology.  

### Jobs executions

An orchestration service starts a new job when the OKP4 blockchain validates a user query.
It fetches data, requests services, and performs various computes from one or several off-chain resources according to the necessary tasks to satisfy the user query.

An orchestration service has to report to the protocol the processes status to ensure service agreement enforcements.

> The final result of the execution is returned to the user’s endpoint, which can visualize the new knowledge through BI components, for example.

Reviewer 3: I think the execution result is defined in the workflow, it may be integrated as a new dataset in the protocol or sent elsewhere outside the control of the protocol if the rules allows it.

Corp A and Corp B inform a maximum of metadata, especially to describe the different available datasets, their structures and the nature of their contents. The portal combines the service execution instructions, the governance rules and the descriptive pieces of information in a Turtle file. Then a transaction adds it to the on-chain ontology.

3. Metadata indicate that resources shared by Corp A and Corp B are compatible with a ML job from Corp C. Corp C also uses the administration portal to submit its training model algorithm, governance rules, and service execution instructions.

4. Individual D parameters its cloud environment with the portal, setting up how an orchestration service should store query results.

5. Individual D uses the portal to request service execution with the ML Workflow from Corp C, using data from Corp A and Corp B. He submits a transaction with a Keplr wallet and pays with $KNOW tokens. The blockchain validates the query (access and execution authorizations).

> 6. The orchestration service listens the event from the blockchain. It recovers access keys, and then executes the workflow training algorithm from Corp C, using Corp A & Corp B data it fetches.

Reviewer 4: ```suggestion
6. The orchestration service listens to the event from the blockchain. It recovers access keys, and then executes the workflow training algorithm from Corp C, using Corp A & Corp B data it fetches.
```
